### PR TITLE
Plane: Qassist switch, logic change and enable for tailsitters

### PR DESCRIPTION
**Pull request by IamPete1**
This work was funded by FDrones

There a three bits to this PR. 
1.  Addition of a switch option to enable, disable and force enable of Qassist functionality. Force enable is Qassist always on, enable allows it to work as normal, disable disables it. Switching to disable while in Qassist does not stop the motors, it must be exited by exceeding the transition airspeed. (ie if your Qassist hovering in FBWA disabling with the switch does not mean you fall out the sky)

2. Qassist logic changed slightly. Currently a airspeed estimate is required to enter Qaissit at all, if there is no airspeed angle based Qaissit does not work. This changes that so you must have airspeed to enter airspeed based Qassist but angle and altitude assist work without.

3. Enables Qassist for tailsitters

### CODE REVIEW

**Comment by kd0aij** [comment]
What are the behavior changes for tailsitters with QASSIST enabled? 
I'm not sure how to test this.

**Comment by IamPete1** [comment]
I was testing on your striker copter tailsitter, in normal flight only the wing motors come on, in Qassist all four fire up. Its less obvious on more traditional tail sitters, in that case its just a change of gains.

**Comment by IamPete1** [comment]
added a fix to disable angle boost for tail sitters in Qassist it makes odd stuff happen

**Comment by IamPete1** [comment]
I think the most important tests are that I have not broken existing behaviour rather than testing new. To test Qassist its best to set the assist airspeed to fbw min and carefully slow down in fbwa. Once the airspeed is bellow it will kick over to 'copter' gains.  Probably a good idea to test the 'copter' gains are OK at that airspeed in Qstab or Qacro first. Or use the new switch to enable and disable assist.

**Comment by kd0aij** [comment]
regarding angle_boost for tailsitters, the recommendation is to disable it completely:
https://discuss.ardupilot.org/t/dual-motor-tailsitters/15302/500
I've just been setting Q_A_ANGLE_BOOST=0  but perhaps it would be better to override it at all times.

**Comment by kd0aij** [comment]
OK, this passes my code inspection (for what it's worth).
It would be nice to see some real flight test logs showing the advantages for using QASSIST and how it compares to a "full" motor mask. I believe @losawing prefers motmx=0xF for his quad tailsitters, but for better endurance we'd like to turn most of the motors off in FW modes.
This would be interesting to test in a tri configuration; there are a couple of RealFlight models here:
https://github.com/ArduPilot/SITL_Models/tree/master/RealFlight/markw/Tailsitters/CopterTailsitters

**Comment by IamPete1** [comment]
This is a demo vid of Qassist and using the switch 
https://youtu.be/2Y9IyxzSvOY

This vid from my first attempt at  adding Qassist probably does a better job of showing the possible improvement if more time is spent tunning 

https://youtu.be/6MIPEbqxRwE

**Comment by IamPete1** [comment]
This has now been tested by Letpi. 

https://discuss.ardupilot.org/t/copter-tailsitters/45514/373?u=iampete

**Comment by IamPete1** [comment]
https://youtu.be/CktDVhvEWJ8

Some testing with a two motor none vectored tailsitter, not such a clear improvement as with other types but still a little better, certainly not worse. I think with some more time spent tuning it could be improved, it does not fly all that well in Qmodes either. I would conclude that if setup correctly Qassist could provide benefit for these vehicle too, certainly not something to turn on by default but good to have the option.

File: ArduPlane/quadplane.cpp [hunk 1]
```diff
@@ -897,7 +903,9 @@ void QuadPlane::hold_stabilize(float throttle_in)
         }
     } else {
         motors->set_desired_spool_state(AP_Motors::DesiredSpoolState::THROTTLE_UNLIMITED);
-        attitude_control->set_throttle_out(throttle_in, true, 0);
+        // tailsitters in forward flight Qassist should not use angle boost
+        bool should_boost = !(is_tailsitter() && !in_vtol_mode() && assisted_flight);
```
Comment (tridge) [inline ArduPlane/quadplane.cpp]: this is much too opaque, not clear what situations that affects

Comment (IamPete1) [reply ArduPlane/quadplane.cpp]: changed to an if, also checking !in_vtol_mode() was redundant to assisted_flight

**Comment by IamPete1** [comment]
Changed this so that Qassist no longer works for none-vectored tailsitters, this is because they go unstable if flown at more than 5 - 10m/s, allowing for some airspeed estimation error margin this means the Qassist speed has to be set so low its pointless.  Also tailsitters such as that have large control surfaces so usually do quite well at low speeds in plane modes in any-case.
Note that if your very keen to try it anyway you can still force enable using the switch.

**Comment by IamPete1** [comment]
another vid of this in action

https://discuss.ardupilot.org/t/copter-tailsitters/45514/475?u=iampete

**Comment by IamPete1** [comment]
had to do quite abit of re-work because we were getting Qassist in forward flight but not in the forward flight part of the transition to VTOL. This fixes that.

**Comment by IamPete1** [comment]
I have just tested normal quadplane Qassist, still working as expected

**Pull request merged**